### ISSUE TITLE
[FireMonkey] [Feature Request] support for @grant

### ISSUE THREAD

**Issue by afontenot** [erosman/support#336]
Currently lines like

    // @grant GM.xmlHttpRequest

in Firemonkey scripts are tagged with the warning `@grant is not processed`. And indeed, scripts appear to have access to this API even with the grant line being provided:

> Scripts will have access to all available APIs and therefore `@grant` is not needed or used.

I think gating access behind `@grant` is a good idea and should be implemented in FM. Certainly, at the end of the day it's the user's responsibility to make sure that a script is trustworthy. But mistakes happen, and in other userscript managers seeing `@grant` is a good sign that a script deserves heightened scrutiny. Some users may even have automatic updates enabled.

Note that along with `@namespace`, `@grant` is one of only two metadata lines that is supported by every userscript manager except FM.

**Comment by erosman** [comment]
As replied in the other topic, `@grant` is not as relevant as it was before.

Other managers behaviour is different in regards to `@grant`.
TM injects into `page` context if `@grant` in `none`.
VM provides GM API in `page` context while GM (doesn't have `page` context), FM & TM don't.
`userScript` context is already heavily sandboxed.

Having `@grant` is just a visual indication. Unless the user inspects the actual code, user can not determine if the script is good or bad.

Granted API by themselves don't increase or decrease user-scripts security in FireMonkey.

**Comment by afontenot** [comment]
I think I can buy this response, as long as there's no way scripts in the `page` context can *ever* get access to these privileged APIs via a broken user script. [This used to be possible](https://groups.google.com/g/greasemonkey-dev/c/kz7NswfEOG0) for any script that used unsafeWindow. Obviously, this question overlaps what we were talking about on the other issue. If they can, `@grant none` becomes nice to have because it limits the damage a broken script can cause.

Either way, I think it might be nice to at least enforce `@grant none`, interpreted as the following:

 * Continue to run in `userScript` context (don't drop to `page` like TM)
 * Do not provide access to any GM api functions

Most scripts would continue to work fine in this mode.

**Comment by erosman** [comment]
> Continue to run in userScript context (don't drop to page like TM)

FM never injects into `page` unless specifically set via `@inject-into` and even then, the user-script will become part of page JS and will be completely isolated from FM & its APIs (no interaction between them any more).

> Do not provide access to any GM api functions

Already the situation in FM

**Comment by afontenot** [comment]
I think you misunderstood the idea.

`@grant none` should do *both* those things simultaneously. Currently, if FM injects into the page, it drops access to GM API functions. Currently, if Tampermonkey sees `@grant none`, it injects into the page.

I'm suggesting that FM support `@grant none`, but unlike TM, don't drop into the page, stay in the secure sandbox. Instead, it will simply block access to all GM API functions. So you get the most secure of both: you're running in the sandboxed context, and you don't have access to GM API. 

If it's possible to give scripts running in the page access to GM APIs via injecting scripts or using unsafeWindow, `@grant none` would provide additional protection against accidentally breaking security in this way. If it's *not* possible to accidentally give page scripts this access (e.g. by overwriting page functions), maybe it doesn't matter. But you mentioned it was possible for content scripts to accidentally provide access to browser APIs, so I'm assuming a similar problem is possible here.

**Comment by erosman** [comment]
> If it's possible to give scripts running in the page access to GM APIs via injecting scripts or using unsafeWindow, 

It is possible to do the same without using unsafeWindow.
It is possible to do the same even with `@grant none`.

It is a false sense of security.

**Comment by afontenot** [comment]
> It is possible to do the same even with `@grant none`.

Hang on. You're telling me that a userscript manager that uses `@grant none` still unavoidably gives access to GM API functions?? How? Even the user scripts themselves aren't supposed to have access to GM API functions when running under `@grant none`, so how is it possible to pass that ability to the page?

Can you provide a example script that contains `@grant none` in which the script itself or the page it's running in gets access to GM API functions in either Greasemonkey or Violentmonkey? That sounds like a major security breach.

**Comment by erosman** [comment]
> You're telling me that a userscript manager that uses @grant none still unavoidably gives access to GM API functions?? 

You misunderstood. User-scripts can do most of what GM APIs do, without the need for GM APIs.

Examples:
What `GM_addStyle()` does, can be done without using  `GM_addStyle()` (that was a very simple API.

What `unsafeWindow` does can be done in mutliple ways:
- `window.wrappedJSObject`
- `window.eval()`
- `<script> DOM`
- etc

openInTab -> window.open
setClipboard -> document.execCommand("copy");
notification _> alert() (not the same but does the job)
.....etc etc etc

**Comment by afontenot** [comment]
> > You're telling me that a userscript manager that uses `@grant none` still unavoidably gives access to GM API functions??
> 
> You misunderstood. User-scripts can do most of what GM APIs do, without the need for GM APIs.

I don't agree that I misunderstood, I took what you said at face value:

> > If it's possible to give scripts running in the page access to GM APIs via injecting scripts or using unsafeWindow,
> 
> It is possible to do the same without using unsafeWindow.
> It is possible to do the same even with `@grant none`.

But never mind that. It seems likely that "most" turns out to be very different than "all". You might as well argue that script managers should deprecate the GM API entirely, since it's not really needed any more. On the contrary:

 * Letting page scripts violate CORS would be a enormous breach of browser security. Therefore page scripts should never have access to `GM.xmlHttpRequest` (or `XMLHttpRequest()` running in `content` scope on Firefox).
 * User scripts are allowed to store resources or data that they want to be kept secure from the page using `GM.getValue` and `GM.setValue`. Scripts in the page should never be able to access this storage area.

If it's true that it's possible for a script to leak access to privileged APIs to the page (and you have made clear that it is), then the obvious value of implementing `@grant` is that it allows script authors to turn off these unneeded APIs. If it's established that implementing `@grant` could provide *some* additional security, is there any reason not to just implement it? (Heck, if as you say GM.\* is hardly useful for anything anymore anyway, why not disable it *by default*, exactly as Greasemonkey does?)

By the way, I'd still be interested in seeing a minimal example of what accidentally letting the page run code in the user script context looks like, if you have one. [Back in 2008](https://groups.google.com/g/greasemonkey-dev/c/kz7NswfEOG0), *any* use of unsafeWindow (or its friends) made this possible, but I believe this hole got plugged since then. Is it now relatively hard to do by mistake? Can it be done without overwriting functions in the page scope?

**Comment by erosman** [comment]
userScripts API was specifically created to combat security issues of contentScripts API which is currently used by GM|VM|TM.

Users are free to choose the user-script manager that best suits their needs.

**Comment by afontenot** [comment]
I'm deeply confused by this conversation up to this point.

> contentScripts API for user-scripts, which is currently used by GM|VM|TM.

As we established on the other thread, the `content` context is not the primary mode of operation for either VM or TM. Most scripts are injected directly into the page for both. I believe [you said this yourself](https://github.com/erosman/support/issues/337#issuecomment-855857124). Maybe I'm being picky on this point, but we appear to be miscommunicating over technicalities.

> The issues related to content context are not relevant to FireMonkey since FM doesn't use it.

Right, that's why I'm not interested in the content context in this thread. I'm only interested in whether or not scripts running in FM can accidentally (or maliciously) give the page access to privileged APIs, **specifically**, the APIs provided by GM.\* such as `GM.xmlHttpRequest`. As my last reply illustrates, there are several privileged APIs which is it very bad to have leaked into the page.

Given this, I think there is a good case to be made that `@grant` support would be beneficial to users of FM. In fact, it might be *more* useful to FM users than GM, precisely because even with `@grant none`, Greasemonkey scripts can still leak *browser APIs* into the page because they're in the `content` context. Scripts running under FM don't even have access to these APIs, so providing the additional protection of `@grant none` (especially by default) would be of great benefit.

**Comment by erosman** [comment]
> As we established on the other thread, the content context is not the primary mode of operation for either VM or TM. Most scripts are injected directly into the page for both. I believe you said this yourself. Maybe I'm being picky on this point, but we appear to be miscommunicating over technicalities.

I can only comment of FireMonkey. Discussing other mangers is not relevant to this support channel.

> I'm only interested in whether or not scripts running in FM can accidentally (or maliciously) give the page access to privileged APIs, specifically, the APIs provided by GM.* such as GM.xmlHttpRequest. 

The implementable of `userScripts` API is more secure for user-scripts than `contentScripts` API. Malicious/poorly-written user-scripts would also have less chance of causing issues.

The function of `GM.xmlHttpRequest` can be achieved by a user-script without using `GM.xmlHttpRequest` by multiple other methods. Therefore there is no new security risk.

`contentScripts` API has access to additional browser API (where as `userScripts` API doesn't) and that is/has been a concern and hence the need to create `userScripts` API.

> Given this, I think there is a good case to be made that @grant support would be beneficial to users of FM.

Average user without programming knowledge will install a user-script whether is has `@grant` or not and would not be able to ascertain the security of a user script.

Users with programming knowledge can check the code for security issue whether is has `@grant` or not.

`@grant none` can only give a false sense of security to users. Malware/poorly-written user-scripts can operate with `@grant none`.

The existence of `@grant` in FireMonkey does not affect its operation (as is the case with other managers). Therefore, in my view the non-implementation of `@grant` in FireMonkey does not create any additional risk.

**Comment by afontenot** [comment]
I'm not denying and don't want to debate that FM is theoretically more secure than other userscript managers because it uses the `userScript` API. The only question at issue is whether a script can be made *more* secure by the use of `@grant none`.

> The function of GM.xmlHttpRequest can be achieved by a user-script without using GM.xmlHttpRequest by multiple other methods. Therefore there is no new security risk.

Please clarify this. It appears to contradict what you said [elsewhere](https://github.com/erosman/support/issues/337#issuecomment-855857124):

> The whole point of having GM XMLHttpRequest is to bypass CORS restriction and nothing else.

This strongly implies that without this API, a script should not be able to bypass CORS restrictions. In other words, it simply isn't true that the function of GM.xmlHttpRequest can be achieved by other methods. In fact the whole point of having this method is to bypass CORS, which (crucially for security) can't be done using native JS methods. (It turns out that content scripts are not subject to CORS as we've discussed on other threads, but lets leave those out of the conversation entirely since FM does not use them.)

As an example of what you're saying, can you provide a simple user script that breaks CORS without using the `GM.*` API (or using a context not subject to CORS)?

> Users with programming knowledge can check the code for security issue whether is has `@grant` or not. `@grant none` can only give a false sense of security to users. Malware/poorly-written user-scripts can operate with `@grant none`.

First of all, the whole question we are discussing here is whether `@grant` is **only** a false sense of security. If it's true that the GM API functions give access to scripts that **must not** ever be given to scripts in the page, then `@grant` is *not* a false sense of security.

Your second point kind of sounds like "if everyone simply writes their code to be free of security issues, there won't be any security issues". I don't feel that that's the right approach. We install firewalls on systems that we believe to be fully up to date and secured. As a script author who's *still* not 100% sure what I need to avoid to not give access to privileged APIs to the page, I feel much more secure knowing that `@grant` is available. At least most of my scripts can't give random domains access to my bank account by breaking CORS - so far as I know.

**Comment by erosman** [comment]
> Please clarify this. It appears to contradict what you said elsewhere:

`GM.XMLHttpReques`t is actually `XMLHttpRequest` sent from a location that has no `origin`. The difference is the `origin`.
Additionally, `GM.XMLHttpRequest` allows setting the `origin` header.

> As an example of what you're saying, can you provide a simple user script that breaks CORS without using the GM.* API (or using a context not subject to CORS)?

Discussion regarding the use of JavaScript `XMLHttpRequest` with regards to `origin` and CORS should be forwarded to JavaScript forums and/or Mozilla.

> First of all, the whole question we are discussing here is whether @grant is only a false sense of security. If it's true that the GM API functions give access to scripts that must not ever be given to scripts in the page, then @grant is not a false sense of security.

To start with, discussion of exploits are off-topic.

The impact of GM API in `content` context and `userScript` context are not the same. Therefore, the impact of `@grant none` in FireMonkey is not the same as extensions using `content` context for user-scripts.

Finally, user-scripts by nature bear security risks and users should only install user-scripts from trusted sources.

**Comment by erosman** [comment]
Let open up the discussion ......

# Sandboxed Context (`content` or `userScript`)

### Does `GM.XMLHttpRequest` contain a security risk since it can bypass CORS?
If the answer is YES, then all user-script managers should remove the API.

### Does `@grant none` prevent XMLHttpRequest CORS?
The CORS security measure is enacted by the target server and depend on the target setting. If the target allow connection from other domains, then it will go through, otherwise not. 
A script on `randomdomain.com` can ONLY do an `XMLHttpRequest` to `mybankaccount.com`, if `mybankaccount.com` allows it.
The existence of `@grant none` has no bearing on it.

While `GM.XMLHttpRequest` might be able to forge `origin`, no-half-decent server will only rely on `origin` for security. CORS is helpful but not the core of server security.

:pushpin: I had already discussed the issue with Mozilla security team before allowing 'origin' to be changed in `GM.XMLHttpRequest` (v1.18), for compatibility with other user-script managers.

### Is `XMLHttpRequest` affected by CORS?
In FireMonkey yes, in others ....... :man_shrugging: 
Users can chose the most secure one.


# Page Context

### Does `GM.XMLHttpRequest` contain a security risk since it can bypass CORS?
Only applies to VM that provides GM API in page context.

### Does `@grant none` prevent XMLHttpRequest CORS?
Only applies to VM.

### Is `XMLHttpRequest` affected by CORS?
It should apply.

**Comment by afontenot** [comment]
I'm fine with this format. I'll answer the same questions you've given. If I don't answer the question, assume I agree or think it's not relevant for what we're discussing here. For one thing, I'll ignore anything about page context (unless I say explicitly otherwise) because I'm only interested in user scripts for our purposes here.

> Does GM.XMLHttpRequest contain a security risk since it can bypass CORS?

Yes, it definitely does. That's why it's an ability only given to trusted user scripts, not arbitrary and potentially dangerous code running in the page. Should it be removed? No, because it's a useful feature that some user scripts require for their operation. It should instead be *limited*, so that it's made available only to those scripts which need it and *explicitly request it*. This is the approach taken by every other userscript manager.

For the rest of your response, I really can't tell if you're suggesting that it would be okay to just disable CORS because sites shouldn't rely on it. Suffice it to say that I strongly disagree. CORS is not really to protect the site, anyway, it's to protect the site's users.

If you really don't see the point of CORS I would suggest reading [this article](https://infosecwriteups.com/zoom-zero-day-4-million-webcams-maybe-an-rce-just-get-them-to-visit-your-website-ac75c83f4ef5?gi=3b289a14d183), about how the webcams of 4 million Zoom users were exposed by developers breaking CORS.

> Does `@grant none` prevent XMLHttpRequest CORS?

I assume the has a typo, and you meant to specify `GM.xmlHttpRequest`. In which case the answer is yes, it does. It prevents a script from using this API to bypass CORS because it does not make this privileged API available to the script. 

If you meant ordinary `XMLHttpRequest`, then the question is not coherent, because ordinary `XMLHttpRequest` is always subject to CORS and has nothing to do with the Greasemonkey API.

> Is XMLHttpRequest affected by CORS?

Yes (in the page context). Referring here to `XMLHttpRequest` not `GM.xmlHttpRequest` - the latter does not respect CORS.

You say yes (in FireMonkey). This is wrong, because the question is once again incoherent. At present *all cross-origin requests from `XMLHttpRequest` are blocked in FireMonkey*, as I pointed out in this bug: https://github.com/erosman/support/issues/340 If *all* requests are blocked, it's not meaningful to ask if they're "blocked by CORS".

Let's grant that in Firemonkey, once someone gets cross-origin requests working, these requests will respect CORS. Fine, that seems like a worthwhile security precaution. That would redouble the urgency of adding support for `@grant` because without grant even scripts that don't need it and insecurely access page objects will have access to CORS violating requests, potentially leaking this to the page.

**Comment by erosman** [comment]
If you make a list of single issues similar to what I posted above, I will do my best to answer them.

**Comment by erosman** [comment]
The following 2 userscripts work the same way in FireMonkey. There is no difference in their code and thus their security i.e. script A is not more secure than script B, and vice versa.
The only difference is the visual `@grant           GM.xmlHttpRequest` in the metadata block.

```js
// ==UserScript==
// @name            Script A
// @match           http://www.example.com/*
// @grant           GM.xmlHttpRequest
// ==/UserScript==

GM.xmlHttpRequest({
  url: 'https://abcd.com/',
  onload: response => console.log(response.responseText)
});
```

```js
// ==UserScript==
// @name            Script B
// @match           http://www.example.com/*
// ==/UserScript==

GM.xmlHttpRequest({
  url: 'https://abcd.com/',
  onload: response => console.log(response.responseText)
});
```

**Comment by Iey4iej3** [comment]
> If you really don't see the point of CORS I would suggest reading [this article](https://infosecwriteups.com/zoom-zero-day-4-million-webcams-maybe-an-rce-just-get-them-to-visit-your-website-ac75c83f4ef5?gi=3b289a14d183), about how the webcams of 4 million Zoom users were exposed by developers breaking CORS.

I don't think that this is essentially due to lack of CORS, although CORS might help. A proof-of-concept looks like this:

1. An insecure daemon is listening on `localhost:1234` which would redirect the stream of the webcam. This kind of daemon is sometimes ad hoc since it is only running on the loopback.
2. A userscript access `localhost:1234` and upload the data to a (malicious) server.

When a CORS check is performed on the insecure daemon, it could prevent a normal `XMLHttpRequest` but not `GM_xmlhttpRequest`.

I don't know what is an ideal solution, but seemingly it is a job of browser to restrict the access of ports of localhosts, in the same way as restricting the access of microphones and webcams.

**Comment by erosman** [comment]
> I don't know what is an ideal solution, but seemingly it is a job of browser to restrict the access of ports of localhosts, in the same way as restricting the access of microphones and webcams.

If a `localhost` has an HTTP server, then it will be treated like any HTTP server.

Nonetheless, the CORS discussion was academic in this topic.

The issue of `@grant` implementation has been addressed in https://github.com/erosman/support/issues/336#issuecomment-858045774

**Comment by Iey4iej3** [comment]
> The issue of `@grant` implementation has been addressed in [#336 (comment)](https://github.com/erosman/support/issues/336#issuecomment-858045774)

If I understand correctly, this code is simply ignoring `@grant`. I tend to believe that OP wants that the second code would be invalid. In some sense, it is similar to force explicit declarations of functions in C - all invoked APIs are listed in the metadata block and otherwise throwing an error.

But I don't know whether it is Firefox which determines what userscript APIs would be accessible to userscript?

**Comment by erosman** [comment]
> If I understand correctly, this code is simply ignoring @grant. 

FireMonkey does not process `@grant` as mentioned in the Help.

> But I don't know whether it is Firefox which determines what userscript APIs would be accessible to userscript?

There are 2 types of API, browser API and GM API.

**Browser API**
- FireMonkey operates in dedicated `userScript` context where no browser APIs are available
- Other userscript managers operate in `content` context where by default some browser APIs are also available

**GM API**
GM APIs are solely provided by the userscript managers. 
- FireMonkey makes all of them available
- Other userscript managers make them available one by one via `@grant`

**Comment by Iey4iej3** [comment]
> FireMonkey does not process `@grant` as mentioned in the Help.

That is why the title of this post is "feature request" - the OP hoped that it would be implemented. To be fair, for me, I don't really care much about this, since indeed there are hundreds of ways to contain malicious codes even if the used APIs are declared, and there are hundreds of ways to obfuscate codes.

> * FireMonkey makes all of them available
> 
> * Other userscript managers make them available one by one via `@grant`

Let me ask one question: is it difficult to implement the restriction on GM API calls for FireMonkey (and other possible addons using userscript context)?

**Comment by erosman** [comment]
> Let me ask one question: is it difficult to implement the restriction on GM API calls for FireMonkey

No.

I did not implement `@grant` because I dont not see a reason for doing so.
In my view, `@grant` only provides a visual indication and nothing else.
